File: DSA-JAVASCRIPT/Data-structures/Array-structure/kadanes_algorithm.js
function maxSumSubArray(arr, size) {

    let currSum = 0, maxSum = Number.MIN_VALUE, startIndex = 0, end = 0, start = 0;

    for (let i = 0; i < size; i++) {
        currSum += arr[i];

        if (currSum > maxSum) {
            maxSum = currSum;
            startIndex = start;
            end = i;
        }
        if (currSum < 0) {
            currSum = 0;
            start = i + 1;
        }
    }
    console.log("Maximum contiguous sum:", maxSum);
    console.log("start index:", startIndex);
    console.log("Endning index:", end);
}

const array = [-2, -3, 4, -1, -2, 1, 5, -3];
sum = maxSumSubArray(array, array.length);

File: DSA-PYTHON/Data-structures/Array-structure/kadanes_algorithm.py
from sys import maxsize


def main():
    array = [5, -4, -2, 6, -1]
    sum, start, end = max_sum_subarray(array)
    print(f"Maximum contiguous sum: {sum}\nStarting index: {start}\nEnding index: {end}")


def max_sum_subarray(arr):
    curr_sum = 0
    max_sum = -maxsize - 1
    start = 0

    for i in arr:
        curr_sum += i

        if (curr_sum > max_sum):
            max_sum = curr_sum
            start_index = start
            end = arr.index(i)

        if (curr_sum < 0):
            curr_sum = 0
            start = arr.index(i) + 1

    return max_sum, start_index, end


main()
